### ISSUE TITLE
[Platform] Hardware test: Test if Container Runtime is OCI compliant

### ISSUE THREAD

**Issue by lixuna** [cncf/cnf-conformance#327]
[Acceptance Criteria](https://github.com/cncf/cnf-conformance/issues/327#issuecomment-668219403)
---

### [Platform]Hardware test: Test if Container Runtime is OCI compliant

**Short description of platform test:** 
- If a container runtime is OCI compliant it's an indicator that the runtime is following cloud native practices. 

**Test Category**
- Hardware

**Proof of Concept Tasks:**

Tasks:
- [ ] Add a quick overview of the type of test you are researching/creating a POC for
- [ ] Investigate potential tools for implementation and document findings, https://hackmd.io/new  (add link to ticket)
- [ ] Select a tool to use, minimal/least effort, and add selection to ticket
- [ ] Add new POC test code
- [ ] Add comment suggesting updates as needed for:
  - [ ] the [test categories markdown](https://github.com/cncf/cnf-conformance/blob/master/TEST-CATEGORIES.md)
  - [ ] the [pseudo code markdown](https://github.com/cncf/cnf-conformance/blob/master/PSEUDO-CODE.md)
  - [ ] slide content updates, LINK_TO_UPDATES
  - [ ] the [README](https://github.com/cncf/cnf-conformance/blob/master/README.md)
- [ ] Tag 1 or more people to peer review

### QA tasks

Dev Review: 
- [ ] walk through A/C
- [ ] do you get the expected result?
- [ ] if yes, 
   - [ ] move to `Needs Peer Review` column  
   - [ ] create Pull Request and follow check list
   - [ ] Assign 1 or more people for peer review
- [ ] if no, document what additional tasks will be needed

Peer review:
- [ ] walk through A/C
- [ ] do you get the expected result?
- [ ] if yes, 
   - [ ] move to `Reviewer Approved` column
   - [ ] Approve pull request
- [ ] if no, 
   - [ ] document what did not go as expected, including error messages and screenshots (if possible)
   - [ ] Add comment to pull request
   - [ ] request changes to pull request

**Comment by lixuna** [comment]
Acceptance Criteria - based on https://github.com/cncf/cnf-conformance/issues/320#issuecomment-665803813
---

### Acceptance Criteria for peer review:
--
### Documentation is updated:
- [ ]  Documentation lists new test name
- [ ]  Test Categories md is updated
- [ ]  USAGE md is updated
- [ ]  Installation instructions is updated

###  Common steps
- [ ]  Find or create a K8s cluster to use
   - Talk with team for access to an existing cluster
   - [Kind K8s clusters](https://github.com/cncf/cnf-conformance/blob/master/KIND-INSTALL.md) will work as well
- [ ]  Set the KUBECONFIG environment variable for accessing the target K8s cluster
- [ ]   As a developer, Pull down latest version of CNF conformance test code from {FEATURE} branch
     - [ ]  Check that {FEATURE} Branch passes all A/C 
     - [ ]  Check that {FEATURE} Branch passes tests in Travis CI
- [ ]   As a peer reviewer, check that PR passes Travis CI tests
     - [ ]  if yes, merge PR to master
     - [ ]  if no, make a comment and request that developer make changes to pass tests 
- [ ]   As a peer reviewer, Pull down latest version of CNF conformance test code from `master` branch
- [ ]   Run `shards install`

### Spec test should pass
- [ ]   Run `crystal spec --tag platform`
- [ ]   I expect to see PASSED for the tests


### Update Prerequistite checker
- N/A

### Update example CNFs cnf-conformance.yml, if needed
- [ ] continue, If there is an update to cnf-conformance.yml
- [ ] for each CNF Example
- [ ] Update the example CNF's cnf-conformance.yml to include the new changes
- [ ] Update the example CNF's README if needed
- [ ] Test the example CNF following it's README

### Update yml validator, if needed
- N/A

### When using the test `cnf-conformance chaos_node_failure` command should PASS test successfully:
- [ ]  Run `./cnf-conformance setup`
- [ ]  Run  `./cnf-conformance sample_coredns_setup`
- [ ]  Run  `./cnf-conformance chaos_node_failure`
- [ ]  I expect to see the output and results from the test
- [ ]  I expect to see PASSED for the test. Example:
         - [ ] "✔️ PASSED:  Cluster recovered after node failure [EMOJI]"

### When using the test `crystal spec --tag `chaos_node_failure` command the sample "good" CNF should PASS and the "bad example CNF"  should FAIL test:
- [ ]   Run `crystal spec --tag volume_hostpath_not_found`
- [ ]   I expect to see the output and results from the test
- [ ]   I expect to see "✔️ PASSED: Cluster recovered after node failure [EMOJI]"
- [ ]   I expect to see "❌ FAILED: Cluster did not recover after node failure [EMOJI]"

### I would expect the CNF to pass when I run the entire test category suite (`resilience`)
- [ ]  Follow common steps above
- [ ]  Run  `./cnf-conformance sample_coredns_setup`
- [ ]  Run  `./cnf-conformance resilience`
- [ ]   I expect to see the output and results from all tests in the test category
- [ ]  I expect to see PASSED for the test and test category 

### I would expect the CNF to pass when I run the entire test suite for platform tests: (`platform`)
- [ ]   Follow common steps above
- [ ]   Run  `./cnf-conformance sample_coredns_setup`
- [ ]   Run  `./cnf-conformance platfrom`
- [ ]   I expect to see the output from the test
- [ ]   I expect to see PASSED for the test

### Screenshots of new test and results in the comments
- [ ]  New test should be added to the points file
- [ ]  New test score should be added to the final score
- [ ] I can see a screenshot of the new test in the comments of this ticket

**Comment by taylor** [comment]
The point of this ticket is not to test the runtime directly which could be extensive and probably will be in multiple tickets. 

To create a list of container runtimes that seem to be cloud native we could look at the runtimes to check for things like:
- list that it follows OCI specs
- list that it follows [Kubernetes Container Runtime Interface (CRI)] runtime spec
- listed in the landscape.cncf.io (potentially with some tag indicating cloud nativeness)

**Comment by lixuna** [comment]
Pass/fail messaging examples: 
- Passes: “Your platform is using containerd which is a compliant runtime”
- Fails: “Unable to find a compliant runtime”

**Comment by lixuna** [comment]
Level of effort in pts: 01,2,3,5,8,13,21 pts: 
---

**Comment by lixuna** [comment]
Note: The team would like to meet for an implementation details meeting to discuss:

- Where will the static list of container runtimes be saved?
- How will the list be written?
- How will the test suite find the platform's selected container runtime tool?
 
and any other items, as needed

**Comment by lixuna** [comment]
Note: 

- Example of a Doc to be updated - https://github.com/cncf/cnf-conformance/blob/bfc7362d0c3dc67bc85d297f96b91c29c1014be3/USAGE.md#hardware-resources-and-scheduling-tests
    - Change "Run all hardware resources and scheduling tests
crystal src/cnf-conformance.cr hardware_affinity"

**Comment by taylor** [comment]
Estimated level of effort: 5pts

- @williscool 5
- @wavell 8
- @denverwilliams 5
- @lixuna 5
- @agentpoyo 5 
- @taylor 8

**Comment by williscool** [comment]
@wavell @denverwilliams getting zero tests when I run `crystal spec --tag platform`

```
A total of 3 warnings were found.
I, [2020-08-24 22:09:15 -07:00 #25601]  INFO -- cnf-conformance: Build Success!


Finished in 35.41 seconds
0 examples, 0 failures, 0 errors, 0 pending
william@Will-LAPTOP-L47SIB7:/mnt/c/Users/william/unison-with-dropbox-symlinked-to-linux/vulk_stuff/cnf-conformance$ crystal spec --tag platform

```

**Comment by williscool** [comment]
test by itself

![image](https://user-images.githubusercontent.com/394931/91126020-e793fc80-e657-11ea-9029-f7e01f329fe5.png)

**Comment by williscool** [comment]
also getting 0 tests with `crystal spec --tag platform:oci_compliant`

**Comment by williscool** [comment]
`./cnf-conformance hardware_and_scheduling` taks not found


```
william@Will-LAPTOP-L47SIB7:/mnt/c/Users/william/unison-with-dropbox-symlinked-to-linux/vulk_stuff/cnf-conformance$ ./cnf-conformance hardware_and_scheduling
lib/sam/src/sam.cr:41:5 in 'invoke'
lib/sam/src/sam.cr:77:7 in 'process_tasks'
lib/sam/src/sam.cr:56:7 in 'help'
src/cnf-conformance.cr:54:1 in '__crystal_main'
../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:105:5 in 'main_user_code'
../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:91:7 in 'main'
../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:114:3 in 'main'
__libc_start_main
_start
???
Task hardware_and_scheduling was not found


```

**Comment by denverwilliams** [comment]
> 25601

Running spec tests against an entire category ```crystal spec --tag platform``` won't work as there aren't any tags for it.

We haven't done this for the existing workload spec tests, and it's not in the A/C for this ticket. But maybe it's something we'll want to add later on? @wavell @taylor

**Comment by denverwilliams** [comment]
> @wavell @denverwilliams getting zero tests when I run `crystal spec --tag platform`
> 
> ```
> A total of 3 warnings were found.
> I, [2020-08-24 22:09:15 -07:00 #25601]  INFO -- cnf-conformance: Build Success!
> 
> 
> Finished in 35.41 seconds
> 0 examples, 0 failures, 0 errors, 0 pending
> william@Will-LAPTOP-L47SIB7:~/cnf-conformance$ crystal spec --tag platform
> ```

Running spec tests against an entire category crystal spec --tag platform won't work as there aren't any tags for it.

We haven't done this for the existing workload spec tests, and it's not in the A/C for this ticket. But maybe it's something we'll want to add later on? @wavell @taylor

**Comment by denverwilliams** [comment]
> also getting 0 tests with `crystal spec --tag platform:oci_compliant`

I've added the missing tag and opened a PR. Once this is merged https://github.com/cncf/cnf-conformance/pull/371 this should work.

**Comment by denverwilliams** [comment]
> `./cnf-conformance hardware_and_scheduling` taks not found
> 
> ```
> william@Will-LAPTOP-L47SIB7:/mnt/c/Users/william/unison-with-dropbox-symlinked-to-linux/vulk_stuff/cnf-conformance$ ./cnf-conformance hardware_and_scheduling
> lib/sam/src/sam.cr:41:5 in 'invoke'
> lib/sam/src/sam.cr:77:7 in 'process_tasks'
> lib/sam/src/sam.cr:56:7 in 'help'
> src/cnf-conformance.cr:54:1 in '__crystal_main'
> ../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:105:5 in 'main_user_code'
> ../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:91:7 in 'main'
> ../../../../../../../home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/crystal/main.cr:114:3 in 'main'
> __libc_start_main
> _start
> ???
> Task hardware_and_scheduling was not found
> ```
This is failing because it needs to have the 'platform' namespace appened to it. So you will need to run ```./cnf-conformance platform:hardware_and_scheduling``` I've updated the A/C to have the correct command. @williscool

**Comment by williscool** [comment]
hmm so @denverwilliams i merged you pr and no matter what I do I can't get

```
crystal spec --tag platform:oci_compliant
```

```
william@Will-LAPTOP-L47SIB7:/mnt/c/Users/william/unison-with-dropbox-symlinked-to-linux/vulk_stuff/cnf-conformance$ crystal spec -v --tag platform:oci_compliant
        In /home/william/.anyenv/envs/crenv/versions/0.35.1/share/crystal/src/logger.cr:124:3

[...]

A total of 3 warnings were found.

Finished in 18.34 seconds
0 examples, 0 failures, 0 errors, 0 pending

```

to find the tests

if I run `crystal spec -v spec/platform/hardware_and_scheduler.cr` on the other hand the test runs but there is no test of the failure case (❌ FAILED: Unable to find a OCI compliant runtime [EMOJI]") that is part of the ac



```
william@Will-LAPTOP-L47SIB7:/mnt/c/Users/william/unison-with-dropbox-symlinked-to-linux/vulk_stuff/cnf-conformance$ crystal spec -v spec/platform/hardware_and_scheduler.cr
[...]
Warning: Deprecated Logger#log. Use `Log` module instead

A total of 3 warnings were found.
Platform
  'oci_compliant' should pass if all runtimes are oci_compliantfind: ‘cnfs/*’: No such file or directory
  'oci_compliant' should pass if all runtimes are oci_compliant

Finished in 21.25 seconds
1 examples, 0 failures, 0 errors, 0 pending

```

**Comment by williscool** [comment]
as far as `./cnf-conformance platform:hardware_and_scheduling`

when I run that you see `Resilience final score: 0 of 15` 

![image](https://user-images.githubusercontent.com/394931/91272754-17fb9980-e731-11ea-9a78-b3a6e616f984.png)

**Comment by williscool** [comment]
also if `crystal spec --tag platform` is out of scope for this ticket can we remove that from the AC?

**Comment by nupejosh** [comment]
@williscool when we run the following command, `./cnf-conformance platform:hardware_and_scheduling` ,  the `0 of 15 tests passed` pictured in the screenshot below is the result of this test not being declared in the points.yml. However there are a few more things that need addressing aside from the points.yml declaration. I believe @denverwilliams is currently working to address these issues 
![Screen Shot 2020-08-26 at 5 44 08 PM](https://user-images.githubusercontent.com/3164983/91364176-c51ef200-e7c3-11ea-8abb-b2bf1fe9aca4.png)

**Comment by williscool** [comment]
👍  ok I'll move back to in progress then

**Comment by wavell** [comment]
./cnf-conformance platform:hardware_and_scheduling is working
![image](https://user-images.githubusercontent.com/29548/91489201-08d23400-e876-11ea-9a0f-70b74bdb1a2d.png)

**Comment by wavell** [comment]
Made a new issue for the platform:resilience totals:
https://github.com/cncf/cnf-conformance/issues/382
